### DISCUSSION TITLE
How do virtual machine platforms work?

### DISCUSSION THREAD

**Discussion by geiseri** [coder/coder#12968] [General] [answered]
I am trying to use the terraform provider for [hive fabric](https://registry.terraform.io/providers/jmullen-hive/hiveio/8.4.2) to provision my virtual machine development environment.  I used the digital ocean template as a starting point.  Everything actually "Just Worked", save for restarting the workspace.  It seems that once the workspace is stopped that virtual machine and coder seem to go off into the weeds and all I can do is delete the workspace.  I assume that it is due to some feature/behavior the provider lacks. Where might I be able to learn more about what features are required by a provider for managing coder workspaces.

**Comment by matifali** [comment]
> It seems that once the workspace is stopped that virtual machine and coder seem to go off into the weeds, and all I can do is delete the workspace

How is Coder deployed? Even if the template is wrong, the Coder should work.

Also, please share your template. And create a help thread at https://discord.gg/coder for more interactive support.

**Reply by geiseri** [reply]
Okay, so I found the answer independently, but it was not the most obvious.  

The line `count = data.coder_workspace.me.start_count` seems to be the magic of if a resource is deleted on stop or not.  

So for example:
```
resource "hiveio_disk" "uv" {
  filename     = "coder-${data.coder_workspace_owner.me.name}-${lower(data.coder_workspace.me.name)}-userdata.qcow2"
  storage_pool = data.hiveio_storage_pool.uv.id
  size         = data.coder_parameter.uv_size.value
  lifecycle {
    ignore_changes = all
  }
}

resource "hiveio_disk" "main" {
  filename       = "coder-${data.coder_workspace_owner.me.name}-${lower(data.coder_workspace.me.name)}.qcow2"
  storage_pool   = data.hiveio_storage_pool.vms.id
  src_storage    = data.hiveio_storage_pool.src.id
  src_filename   = "ubuntu-${data.coder_parameter.cloud_image.value}.img"
  size           = 80
  format         = "qcow2"
  count          = data.coder_workspace.me.start_count
}
```

In this case, when the workspace is stopped the `main` disk resource is deleted and then recreated when the workspace starts.
On the other hand `uv` will be kept around until the workspace is deleted. 

Not sure if this is the best way to do it but it worked for me.  @matifali is this a valid approach?

**Reply by matifali** [reply, marked as answer]
Yes you are right @geiseri. `count = data.coder_workspace.me.start_count` should only be used with resources you want to make ephemeral. i.e. containers etc.
